diff -urpN tor-0.2.3.25.orig/src/or/or.h tor-0.2.3.25/src/or/or.h
--- tor-0.2.3.25.orig/src/or/or.h	2012-11-19 13:24:38.000000000 -0800
+++ tor-0.2.3.25/src/or/or.h	2013-06-11 23:02:33.507804078 -0700
@@ -3746,7 +3746,7 @@ struct socks_request_t {
 /********************************* circuitbuild.c **********************/
 
 /** How many hops does a general-purpose circuit have by default? */
-#define DEFAULT_ROUTE_LEN 3
+#define DEFAULT_ROUTE_LEN 2
 
 /* Circuit Build Timeout "public" structures. */
 
